dataRetriever/dataPool/headersCache/headersNonceCache.go

```diff
@@ -0,0 +1,252 @@
+		hnc.hdrsCounter.decrement(hdrInfo.headerShardId)
+
+		if len(hdrListD.headerList) == 0 {
+			hnc.hdrNonceCache[hdrInfo.headerShardId].removeElement(hdrInfo.headerNonce)
+			return
+		}
+
+		hnc.hdrNonceCache[hdrInfo.headerShardId].addElement(hdrInfo.headerNonce, hdrListD)
+		return
+	}
+}
+
+func (hnc *headersNonceCache) removeHeaderNonceByNonceAndShardId(hdrNonce uint64, shardId uint32) int {
+	_, ok := hnc.hdrNonceCache[shardId]
+	if !ok {
+		return 0
+	}
+
+	hdrListD, ok := hnc.hdrNonceCache[shardId].getHeadersDetailsListFromSMap(hdrNonce)
+	if !ok {
+		return 0
+	}
+
+	hdrsHashes := make([][]byte, 0)
+	for _, hdrDetails := range hdrListD.headerList {
+		hdrsHashes = append(hdrsHashes, hdrDetails.headerHash)
+		hnc.hdrsCounter.decrement(shardId)
+	}
+
+	hnc.hdrNonceCache[shardId].removeElement(hdrNonce)
+
+	//remove elements from hashes map
+	for _, hash := range hdrsHashes {
+		hnc.headersByHash.deleteElement(hash)
+	}
+
+	return len(hdrsHashes)
+}
+
+func (hnc *headersNonceCache) getHeadersByNonceAndShardId(hdrNonce uint64, shardId uint32) ([]headerDetails, bool) {
+	hnc.mutHeadersMap.RLock()
+	defer hnc.mutHeadersMap.RUnlock()
+
+	_, ok := hnc.hdrNonceCache[shardId]
+	if !ok {
+		return nil, false
+	}
+
+	headersList, ok := hnc.hdrNonceCache[shardId].getHeadersDetailsListFromSMap(hdrNonce)
+	if !ok {
+		return nil, false
+	}
+
+	return headersList.headerList, true
+}
+
+func (hnc *headersNonceCache) getNumHeaderFromCache(shardId uint32) int64 {
+	return hnc.hdrsCounter.getNumHeaderFromCache(shardId)
+}
+
+func (hnc *headersNonceCache) lruEviction(shardId uint32) {
+	nonces := hnc.hdrNonceCache[shardId].getNoncesTimestampSorted()
+
+	var numHashes int
+	for i := 0; i < hnc.numHeadersToRemove && i < len(nonces); i++ {
+		numHashes += hnc.removeHeaderNonceByNonceAndShardId(nonces[i], shardId)
+
+		if numHashes >= hnc.numHeadersToRemove {
+			break
+		}
+	}
+}
+
+func (hnc *headersNonceCache) removeHeaderByHash(hash []byte) {
+	hnc.mutHeadersMap.Lock()
+	defer hnc.mutHeadersMap.Unlock()
+
+	info, ok := hnc.headersByHash.getElement(hash)
+	if !ok {
+		return
+	}
+
+	//remove header from first map
+	hnc.removeHeaderNonceCache(info, hash)
+	//remove header from second map
+	hnc.headersByHash.deleteElement(hash)
+}
+
+func (hnc *headersNonceCache) getHeaderByHash(hash []byte) (data.HeaderHandler, error) {
+	info, ok := hnc.headersByHash.getElement(hash)
+	if !ok {
+		return nil, ErrHeaderNotFound
+	}
+
+	headersList, ok := hnc.getHeadersByNonceAndShardId(info.headerNonce, info.headerShardId)
+	if !ok {
+		return nil, ErrHeaderNotFound
+	}
+
+	for _, hdrDetails := range headersList {
+		if bytes.Equal(hash, hdrDetails.headerHash) {
+			return hdrDetails.header, nil
+		}
+	}
+
+	return nil, ErrHeaderNotFound
+}
+
+func (hnc *headersNonceCache) keys(shardId uint32) []uint64 {
+	shardMap := hnc.getShardMap(shardId)
+
+	return shardMap.keys()
+}
+
+func (hnc *headersNonceCache) tryToDoEviction(hdrShardId uint32) {
+	hnc.mutHeadersMap.Lock()
```

Comment:
So there is a global lock (disregarding shard) for eviction. This global lock is also activates by `addHeaderInNonceCache`, since that calls eviction. This lock is short, indeed, but can you check the `eviction is needed` condition in a read lock perhaps?

Or simply remove all the secondary locks from secondary structures and all the read/write locks, and just keep a single global lock.

Do not use channels for this kind of sync purposes. It's hard to read and debug. If you need one lock for each shard, then why not a map of locks? Or array of locks. Take a look on this for inspiration: https://github.com/orcaman/concurrent-map/blob/master/concurrent_map.go

Reply:
done. headers cache use just one mutex.